### PR TITLE
validate: increase OS validation for special cases

### PR DESCRIPTION
**Pull request by q384566678**
For some of the [linux system](https://github.com/opencontainers/runtime-spec/blame/master/config.md#L145) under the special value of the verification, should increase the decision on the `os`.

Signed-off-by: zhouhao <zhouhao@cn.fujitsu.com>

### CODE REVIEW

File: validate/validate.go [hunk 1]
```diff
@@ -256,28 +256,30 @@ func (v *Validator) CheckProcess() (msgs []string) {
 		}
 	}
 
-	for _, capability := range process.Capabilities {
-		if err := CapValid(capability, v.HostSpecific); err != nil {
-			msgs = append(msgs, fmt.Sprintf("capability %q is not valid, man capabilities(7)", capability))
+	if v.spec.Platform.OS == "linux" {
+		for _, capability := range process.Capabilities {
```
Comment (wking) [inline validate/validate.go]: `process.capabilities` is no longer Linux-specific since opencontainers/runtime-spec#673 (which [removed this line][1]).  I'm not clear on how the capabilities check is supposed to work on other OSes, but unless the spec changes I don't think we can have a Linux-only check here.

[1]: https://github.com/opencontainers/runtime-spec/pull/673/files#diff-c9c91c29b41257aea3a3403cc606ad99L132

File: validate/validate.go [hunk 2]
```diff
@@ -256,28 +256,30 @@ func (v *Validator) CheckProcess() (msgs []string) {
...
-	}
 
-	for index, rlimit := range process.Rlimits {
-		if err := rlimitValid(rlimit); err != nil {
-			msgs = append(msgs, err.Error())
-		}
-		for i := index + 1; i < len(process.Rlimits); i++ {
-			if process.Rlimits[index].Type == process.Rlimits[i].Type {
-				msgs = append(msgs, fmt.Sprintf("rlimit can not contain the same type %q.", process.Rlimits[index].Type))
+		for index, rlimit := range process.Rlimits {
```
Comment (wking) [inline validate/validate.go]: `process.rlimits` is also no longer Linux-specific since opencontainers/runtie-spec#673, so this shouldn't get moved into a Linux-specific section.

File: validate/validate.go (hunk 1 above)
Comment (q384566678) [inline validate/validate.go]: But the validation of `capabilities` is compared with `CAPABILITIES (7)`,  [CAPABILITIES(7)](http://man7.org/linux/man-pages/man7/capabilities.7.html) is the value under linux.

Comment (wking) [inline validate/validate.go]: So we should add checks for capabilities on other OSes.  Or move the Linux-specific cap validation into a `CheckLinuxCapabilities` and raise not-implemented-yet errors if someone tries to validate a non-Linux config.  But silently ignoring `process.capabilities` in non-Linux configs will give users a false sense of security.

Comment (q384566678) [inline validate/validate.go]: But it is not clear the value of other systems, it can not be verified.
My idea is to verify the linux system under the capabilities of the special value, not the entire capabilities field validation, so is not the integration of these tests into a function will be better?

Comment (wking) [inline validate/validate.go]: > But it is not clear the value of other systems, it can not be verified.

I agree, and think somebody with more weight than me should bring this up among the runtime-spec maintainers before they cut 1.0 (I [tried][1], but didn't have much success).

> My idea is to verify the linux system under the capabilities of the special value, not the entire capabilities field validation…

That's fine, but I don't think we should be silently passing on non-Linux OSes.  Does [moving the Linux-specific cap validation into `CheckLinuxCapabilities` and raising not-implemented-yet errors for non-Linux configs][2] sound like a bad idea to you?

[1]: https://github.com/opencontainers/runtime-spec/pull/673/files#r99353136
[2]: https://github.com/opencontainers/runtime-tools/pull/350#discussion_r107457088

**Comment by q384566678** [comment]
@wking What do you think of https://github.com/opencontainers/runtime-tools/pull/350/commits/15804036a2282a41aaaafde2b9aea8b2af77dde1?

File: validate/validate.go [hunk 3]
```diff
@@ -256,29 +256,52 @@ func (v *Validator) CheckProcess() (msgs []string) {
...
+
+func (v *Validator) CheckCapablities() (msgs []string) {
+	if v.spec.Platform.OS == "linux" {
+		for _, capability := range v.spec.Process.Capabilities {
+			if err := CapValid(capability, v.HostSpecific); err != nil {
+				msgs = append(msgs, fmt.Sprintf("capability %q is not valid, man capabilities(7)", capability))
 			}
 		}
+	} else {
+		logrus.Warnf("OS %q has not yet have a special value for capabilities", v.spec.Platform.OS)
```
Comment (wking) [inline validate/validate.go]: Maybe reword to:

> process.capabilities validation not yet implemented for OS %q

**Comment by q384566678** [comment]
@mrunalp @liangchenye @Mashimiao PTAL

File: validate/validate.go [hunk 4]
```diff
@@ -273,47 +273,71 @@ func (v *Validator) CheckProcess() (msgs []string) {
...
+	process := v.spec.Process
+	if v.spec.Platform.OS == "linux" {
+		for index, rlimit := range process.Rlimits {
+			if err := rlimitValid(rlimit); err != nil {
+				msgs = append(msgs, err.Error())
+			}
+			for i := index + 1; i < len(process.Rlimits); i++ {
+				if process.Rlimits[index].Type == process.Rlimits[i].Type {
+					msgs = append(msgs, fmt.Sprintf("rlimit can not contain the same type %q.", process.Rlimits[index].Type))
+				}
```
Comment (Mashimiao) [inline validate/validate.go]: duplicate rlimits check should work for all OSs, not just Linux

Comment (q384566678) [inline validate/validate.go]: updated, PTAL.

**Comment by Mashimiao** [comment]
LGTM

**Comment by liangchenye** [comment]
LGTM
